### ISSUE TITLE
fix: Flutter 3.16 make some goldens fail on CI platform

### ISSUE THREAD

**Issue by orevial** [Betterment/alchemist#104]
### Is there an existing issue for this?

- [X] I have searched the existing issues.

### Version

0.7.0

### Description

Since recent Flutter upgrade to 3.16 I now have issues after regenerating goldens on my machine, some of them fail on CI due to slight differences. 

My process:  

1. I run the tests and update the goldens on my MacBook Pro using default config `PlatformGoldensConfig(enabled: true)`. All my golden files are generated correctly.
2. If I run a test locally on my MacBook Pro without updating the goldens, everything is fine, tests all succeed.
3. When I push to my CI (running Ubuntu) and goldens tests are run using config `PlatformGoldensConfig(enabled: false)`, then I get a lot (but not all) golden tests failing with a very slight diff percentage.

Now I don't understand what is happening but what I can reproduce for sure is that all golden tests that contain multiple scenarii and hence multiple lines or columns, FAIL. 

## Multiple columns goldens

Take this golden file as an example that contains a dumb container with a border: 

![simple-container-one-column](https://github.com/Betterment/alchemist/assets/5528566/e58c32a6-a835-40fb-be45-af86a4e80306)

Code: 

```dart
  goldenTest(
    'Simple container',
    fileName: 'simple-container-one-column',
    builder: () => GoldenTestGroup(
      columns: 1,
      scenarioConstraints: const BoxConstraints(
        minWidth: 100,
        maxWidth: 100,
        minHeight: 50,
        maxHeight: 50,
      ),
      children: [
        GoldenTestScenario(
          name: '100x50 container with blue border',
          child: _separator(Colors.blue),
        ),
      ],
    ),
  );
```

✅ This golden file is correctly generated and asserted during CI tests.

Now take the same test that just duplicates previous scenario:

```dart
  goldenTest(
    'Simple container, two columns',
    fileName: 'simple-containers-two-columns',
    builder: () => GoldenTestGroup(
      columns: 2,
      scenarioConstraints: const BoxConstraints(
        minWidth: 100,
        maxWidth: 100,
        minHeight: 50,
        maxHeight: 50,
      ),
      children: [
        GoldenTestScenario(
          name: '100x50 container with blue border',
          child: _separator(Colors.blue),
        ),
        GoldenTestScenario(
          name: '100x50 container with blue border',
          child: _separator(Colors.blue),
        ),
      ],
    ),
  );
```

✅On my machine, everything runs as expected: 

![simple-containers-two-columns](https://github.com/Betterment/alchemist/assets/5528566/005559c3-5d23-4604-bbc7-c8de2816a7c1)

❌ But this time on the CI, I have a slight difference showing at t

he left border of my container for the second scenario (second column) as shown in the following golden failure masked & isolated diffs: 

![simple-containers-two-columns_maskedDiff](https://github.com/Betterment/alchemist/assets/5528566/d629e60f-6b9a-43b2-afa8-4a2e95d01efa)

![simple-containers-two-columns_isolatedDiff](https://github.com/Betterment/alchemist/assets/5528566/30e4d966-41cd-47ef-bccf-14db9f1a6be7)

## Multiple lines goldens

Same thing happens for goldens tests that contain a single column but multiple scenarii. 

Take this example that runs smoothly on my machine : 

![checkboxes](https://github.com/Betterment/alchemist/assets/5528566/c72768d0-7df5-4672-9bdb-9f378af7ecf8)

❌ Now on the CI, I would get a very subtle one pixel diff at the end of the scenarii separator.

Here is the isolated diff:

![checkboxes_isolatedDiff](https://github.com/Betterment/alchemist/assets/5528566/459487aa-aadc-4614-b63a-41a46857a1cd)

I've pointed the actual 1 pixel difference with a big red arrow because the diff is very subtle 😅

## Conclusion

What is happening here ? I would expect the CI test, that explicitly doesn't display fonts, would work well. 

There seems to be an issue with how alchemist package separate the different scenarii in multiple lines or columns after upgrading to Flutter 3.16.

### Steps to reproduce

1. Add the following code:
  ```dart
  import 'package:alchemist/alchemist.dart';
  import 'package:flutter/material.dart';
  
  Widget _separator(
    Color color, [
    String? text,
  ]) =>
      Container(
        decoration: BoxDecoration(
          border: Border.all(color: color, width: 2),
        ),
        height: 25,
        child: text != null
            ? Text(
                text,
                textAlign: TextAlign.center,
              )
            : null,
      );
  
  void main() {
    goldenTest(
      'Simple container',
      fileName: 'simple-container-one-column',
      builder: () => GoldenTestGroup(
        columns: 1,
        scenarioConstraints: const BoxConstraints(
          minWidth: 100,
          maxWidth: 100,
          minHeight: 50,
          maxHeight: 50,
        ),
        children: [
          GoldenTestScenario(
            name: '100x50 container with blue border',
            child: _separator(Colors.blue),
          ),
        ],
      ),
    );
  
    goldenTest(
      'Simple container, two columns',
      fileName: 'simple-containers-two-columns',
      builder: () => GoldenTestGroup(
        columns: 2,
        scenarioConstraints: const BoxConstraints(
          minWidth: 100,
          maxWidth: 100,
          minHeight: 50,
          maxHeight: 50,
        ),
        children: [
          GoldenTestScenario(
            name: '100x50 container with blue border',
            child: _separator(Colors.blue),
          ),
          GoldenTestScenario(
            name: '100x50 container with blue border',
            child: _separator(Colors.blue),
          ),
        ],
      ),
    );
  }
  ```
2. Generate the golden tests (`flutter test update-goldens`) on a first platform (e.g. a MacBook Pro or Windows machine)
3. Run the tests on a CI platform (e.g. Ubuntu)
4. Check for failures

### Expected behavior

Golden tests should pass and we shouldn't have diffs that are introduced by alchemist library itself (on its separators).

**Comment by Arthurius** [comment]
Any news on this issue? It's really annoying that some tests keep failing because of this

**Comment by vanlooverenkoen** [comment]
@Kirpal or @jeroen-meijer sorry that I tag you both personally, but I feel like this is a crucial bug that should be fixed for Flutter 3.16.x, any idea if someone is already taking a look at this issue?

**Comment by vanlooverenkoen** [comment]
Updating the goldens on Linux results in changed images. But running the tests again on our dev machine (macos) they fail -> which makes sense, because previously our goldens were built using macos. It seems that there were some changes on the goldentests itself.

**Comment by vanlooverenkoen** [comment]
Might be related to: https://github.com/Betterment/alchemist/issues/104

**Comment by vanlooverenkoen** [comment]
Might be related to https://github.com/flutter/flutter/issues/51344

**Comment by vanlooverenkoen** [comment]
Just to be 100% sure @orevial your CI is also Linux right? (a different OS to your dev machine?)

**Comment by orevial** [comment]
> Just to be 100% sure @orevial your CI is also Linux right? (a different OS to your dev machine?)

Yes my CI is a Ubuntu machine!

**Comment by vanlooverenkoen** [comment]
I feel like flutter or alchemist, I'm currently not sure if it is a flutter issue or an alchemist issue, renders on half pixels and that seem to be different on linux & macos. 
master (macos):
<img width="952" alt="image" src="https://github.com/Betterment/alchemist/assets/21172855/0d042983-ab77-497b-bcf9-a4bd9836e2bd">
test (linux):
<img width="949" alt="image" src="https://github.com/Betterment/alchemist/assets/21172855/c589dda8-6214-4627-8c14-7735f7024520">

That also explains why some tests are failing and some are succeeding. I am trying some things in this PR: https://github.com/vanlooverenkoen/alchemist/pull/1

This log for example shows that not all tests are failing: https://github.com/vanlooverenkoen/alchemist/actions/runs/7500793916/job/20420196325

**Comment by vanlooverenkoen** [comment]
@btrautmann can you help us with this issue? I feel like whenever something that should be rendered is 1 pixel or smaller it results in failures on CI. As wel as disabling shadow rendering. Instead a border is rendered somewhere (I have not been able to find where exactly this is done) but that also results in differences. (The border happens to be 1pixel or smaller)

**Comment by vanlooverenkoen** [comment]
This some test that only uses flutter_test api's also fails on ci

```dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('flutter smoke test', () {
    testWidgets('Flutter golden test', (WidgetTester tester) async {
      final mementoDebugDisableShadows = debugDisableShadows;
      // Enable shadows
      debugDisableShadows = false;
      await tester.pumpWidget(
        MaterialApp(
          home: Center(
            child: ElevatedButton(
              onPressed: () {},
              onLongPress: () {},
              child: const Text('button'),
            ),
          ),
        ),
      );
      await tester.pumpAndSettle();
      await expectLater(
        find.byType(ElevatedButton),
        matchesGoldenFile('goldens/flutter_golden_test.png'),
      );

      // Set the flag back to normal
      debugDisableShadows = mementoDebugDisableShadows;
    });
  });
}
```

So it should be a Flutter 3.16 issue.

**Comment by vanlooverenkoen** [comment]
@Piinks do you have an idea if something changed in 3.16 that could affect this?

**Comment by vanlooverenkoen** [comment]
@orevial do you have any tests that are flutter_test only (so no alchemist) where you have the same issues? I can't seem to find anything in alchemist that would fix/cause this bug. But on the other hand I can't find any case where it happens with flutter_test only. 🤯

**Comment by orevial** [comment]
> @orevial do you have any tests that are flutter_test only (so no alchemist) where you have the same issues? I can't seem to find anything in alchemist that would fix/cause this bug. But on the other hand I can't find any case where it happens with flutter_test only. 🤯

That was exactly where I landed when I tried to investigate this issue within alchemist package... 

To answer your question: not really because my non-alchemist golden tests run with real fonts enabled, which are known to render differently on different platforms (what I mean is that they will always fail anyway...). I will try and remove the font loading and keep the default Ahem font to see whether my tests are different between my macOs and Ubuntu platforms...

**Comment by vanlooverenkoen** [comment]
@orevial thanks, Yeah I have no idea where to go from here. And I think I have tagged all people who might know how to fix the issue. I guess at some point there will be more people who need 3.16 support. (which is already out for 2 months already)

**Comment by orevial** [comment]
Agreed, thanks for your help ! 
My "workaround" in the meantime is to always run the golden tests on the same platform: 

- Locally I now run my golden tests in a dockerized Ubuntu with Flutter 3.16 installed (With commands like: `docker exec $container_name /bin/bash -c "cd $package_path;flutter pub get;flutter test $update_goldens --tags golden"`)  
- My CI uses Github Action which runs the tests on an Ubuntu machine that also have  Flutter 3.16 installed

It's not great for local performances because running tests in a Docker image take twice the time as running them directly with Flutter CLI, but at least it make tests repeatable and robust...

**Comment by vanlooverenkoen** [comment]
Yeah indeed. I think we will move our tolerance up until there is a fix. 

@betterment & @verrygoodventures if there is something I need to do or can test let me know I am happy to help. But I am just stuck right now.

**Comment by vanlooverenkoen** [comment]
Hmm yeah indeed. I guess I will wait on a proper solution from the alchemist team itself.

**Comment by maxAtIppen** [comment]
Having the same issue.

Flutter: 3.22.2
CI: Ubuntu
alchemist: 0.7.0

**Comment by btrautmann** [comment]
> @btrautmann were you able to reproduce this?

Yes, as mentioned in https://github.com/Betterment/alchemist/issues/103#issuecomment-2341208297 there are small diffs even when disabling `useMaterial3`. This is almost certainly to underlying Flutter changes, which is why I feel comfortable re-generating Alchemist's smoke test goldens.

There is only _one_ diff that I don't _really_ understand, but again, nothing changed on Alchemist's part as in the upgrade to 3.16.0, so this is likely also related to Flutter. I think a matrix of tests that runs daily will help us to better isolate when these changes occur. If we start failing we can look at what's landed on the beta channel and the culprit will likely be more obvious. Running on each commit _to_ beta may be even better, just not sure how feasible that is. 

![blocked_text_image_reference_isolatedDiff](https://github.com/user-attachments/assets/58627411-eec7-4fe3-ab58-32999724ef66)
![blocked_text_image_reference_maskedDiff](https://github.com/user-attachments/assets/024205fc-917f-407a-ace6-2b160fd88a4d)
![blocked_text_image_reference_masterImage](https://github.com/user-attachments/assets/e352a930-13a0-4c94-ad71-62df321f661d)
![blocked_text_image_reference_testImage](https://github.com/user-attachments/assets/712c20f2-a903-4a28-904c-1809e4e56964)

**Comment by btrautmann** [comment]
@vanlooverenkoen you can see the diff of re-generated goldens in https://github.com/Betterment/alchemist/pull/118 if you want to take a peek.

**Comment by btrautmann** [comment]
**Update:** We recently released 0.8.0 which has a minimum version of 3.16.0. As part of that (and as mentioned in https://github.com/Betterment/alchemist/issues/103#issuecomment-2341208297), we re-generated the golden images to be compatible with the latest Flutter changes. The goal is to do that as seldomly as necessary in the future, but here it was a necessity. I've opened https://github.com/Betterment/alchemist/pull/123 as a way of getting more advanced notice of these types of breaks in the future, but will need to find a way to decouple alchemist from the Material changes that keep having downstream impacts before we can enable that workflow, as mentioned in https://github.com/Betterment/alchemist/issues/120.

I'm going to close this issue in the meantime.
